stdlib/lazy.mli

```diff
@@ -67,6 +68,23 @@ external force : 'a t -> 'a = "%lazy_force"
    recursively.
 *)
 
+val is_val : 'a t -> bool
+(** [is_val x] returns [true] if [x] has already been forced and
+    did not raise an exception.
+    @since 4.00.0 *)
+
+val map : ('a -> 'b) -> 'a t -> 'b t
+(** [map f x] returns a suspension that, when forced,
```

Comment:
Another useful variant would apply `f` eagerly if the suspension had already been forced (particularly useful when `f` is very cheap, to avoid building a useless closure).  One could even argue that this function would be a more natural `map` over lazy values, seen as (mutable) sum types with two constructors (forced value or lazy thunk).  Should we provide both variants (or just one function with an optional flag)?  I'm not strongly asking for it, just wanted to raise the point in case this affects the naming choice.

Reply:
I think this optimization makes sense (but in the same function, no need for a separate function or an extra argument). The docstring should be amended to indicate that `f` could be executed right away if the lazy has already been forced.

Reply:
This is an interesting optimization but its semantics is weird, so it should definitely not be the default behavior.  By "weird semantics", I mean it's neither lazy nor strict.  The lazy behavior is
```
let map f x = lazy (f (force x))
```
as in this PR, and the strict behavior is
```
let map f x = f (force x)
```
In by-the-book functional programming, there is no other option!  The optimized map proposed by @alainfrisch can only be defined if primitives like `Lazy.is_val` are provided, which again is not by-the-book functional programming.

Reply:
I agree with @xavierleroy that this definitely shouldn't be the default. If `f` is expensive and the resulting lazy is not forced, then the optimisation can do unbounded unnecessary work, precisely what the user is using `Lazy` to avoid.

> a more natural map over lazy values, seen as (mutable) sum types with two constructors (forced value or lazy thunk).

This is indeed the natural map over sums with two constructors, but I see that as an implementation detail of Lazy rather than a spec. The natural map over computations-evaluated-only-if-forced is @gasche's, because it has:
```ocaml
    map f (lazy e) == lazy (f e)
```
where `==` is observational equivalence (assuming you don't get to observe with the "advanced" parts of Lazy). By constrast, the optimised sometimes-eager version:
```ocaml
    ignore (map print_int (lazy (1 + 1)))
```
makes it observable whether the compiler has performed constant-folding.